### PR TITLE
Exclude padding from RL trainer diagnostics, and fix three related metrics

### LINKED ISSUE #6809: Metrics-only inaccuracies in RL trainers: unmasked stats (OnlineDPO, PPO), wrong XPO logps metric, GRPO policy_loss only logged with entropy bonus
### Description

A batch of metrics/logging inaccuracies in the RL trainers, found during a code audit at f72460224c5477e65031b2363f54368699105182. None of these affect the training loss — all loss paths are correctly masked — but they skew the reported diagnostics that people use to monitor runs. Bundling them since each is small; happy to split if preferred.

### 1. Online DPO: KL/entropy/reward stats include padding positions

In `trl/experimental/online_dpo/online_dpo_trainer.py::training_step`, the loss path masks padding (`cr_logprobs * ~cr_padding_mask`), but the stats block uses the raw unmasked tensors:

```python
kl = logprobs - ref_logprobs
mean_kl = kl.sum(1).mean()
...
non_score_reward = (-self.beta * kl).sum(1)
...
mean_entropy = -logprobs.sum(1).mean()
```

`logprobs`/`ref_logprobs` are `(B, T)` over the padded completion width, so `objective/kl`, `objective/non_score_reward`, `objective/rlhf_reward`, and `objective/entropy` all sum garbage logprobs at post-EOS padding positions. The analogous PPO entropy issue was fixed in #6121; the Online DPO copy wasn't. XPO and Nash-MD mask these stats correctly.

### 2. PPO: remaining unmasked diagnostics

#6121 fixed `objective/entropy`, but in `trl/experimental/ppo/ppo_trainer.py` the optimization-loop diagnostics still average over all positions including padding:

```python
prob_dist = torch.nn.functional.softmax(logits, dim=-1)
entropy = torch.logsumexp(logits, dim=-1) - torch.sum(prob_dist * logits, dim=-1)
approxkl = 0.5 * (logprobs_diff**2).mean()
...
entropy_stats[...] = entropy.mean()
ratio_stats[...] = ratio.mean()
```

`policy/approxkl_avg`, `policy/entropy_avg`, and `policy/ratio_avg` should use `masked_mean(..., ~padding_mask[micro_batch_inds])` like the neighboring `pg_loss`/`pg_clipfrac` already do.

### 3. XPO: `logps/chosen` / `logps/rejected` report policy + reference logps

```python
# trl/experimental/xpo/xpo_trainer.py
self.stats["logps/chosen"].append(gather_mean(chosen_model_logprobs.mean() + chosen_ref_logprobs.mean()))
self.stats["logps/rejected"].append(gather_mean(rejected_model_logprobs.mean() + rejected_ref_logprobs.mean()))
```

The sum of policy and reference logps isn't a meaningful quantity; Online DPO logs the policy logps. Presumably `+` should not be there (or the intent was the log-ratio, which is already logged via rewards).

### 4. GRPO: `policy_loss` metric only logged when the entropy bonus is enabled

In `trl/trainer/grpo_trainer.py::_compute_loss`, every loss-type branch computes `policy_loss = loss.detach()`, but the only append is inside the entropy-bonus block:

```python
if self._entropy_bonus_enabled:
    ...
    self._metrics[mode]["policy_loss"].append(self.accelerator.gather(policy_loss).nanmean().item())
```

So the `policy_loss` metric silently disappears for standard runs (no entropy bonus), even though it's computed unconditionally. The append presumably belongs outside the `if`.

I'm happy to submit a PR.

### Reproduction

Code inspection; each item follows directly from the quoted excerpts.

### System Info

Found by code inspection (AI-assisted audit) at f72460224c5477e65031b2363f54368699105182.

- Platform: macOS-26.5.1-arm64-arm-64bit
- Python version: 3.12.7
- TRL version: 1.5.1+f724602
- PyTorch version: 2.7.1
- Transformers version: 5.8.1
- Accelerate version: 1.8.1

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

### PR DESCRIPTION
**Pull request by behroozazarkhalili**
Online DPO included padding in metrics, XPO combined policy and reference log probabilities, and GRPO omitted `policy_loss` on some paths. I corrected those metrics and tested their values against independent calculations, with the selected trainer tests passing except for Liger tests blocked by missing GPU support. This overlaps #6856, so whichever merges second must carry the metric change into the GSPO-token copy.
Resolves #6809.

---

> [!NOTE]
> **Low Risk**
> Changes affect diagnostics and documentation only; training losses are not modified, with tests pinning the corrected metric values.
> 
> **Overview**
> Fixes misleading **logged metrics** in several RL trainers; **optimization loss paths are unchanged**.
> 
> **Online DPO** now masks padded completion tokens when computing `objective/kl`, `objective/entropy`, and derived objectives (`non_score_reward`, `rlhf_reward`), matching the masked loss path instead of summing post-EOS padding log-probs.
> 
> **XPO** `logps/chosen` and `logps/rejected` now record **policy** log-probability sums only, not policy plus reference.
> 
> **GRPO** (and the experimental **GSPO-token** copy) always appends `policy_loss` to metrics—not only when an entropy bonus is enabled—and the **Liger** branch logs the same quantity at the same normalization point as the non-Liger path (with loss-type-specific scaling documented in `grpo_trainer.md`).
> 
> Tests add numeric oracles for Online DPO padding and XPO logps, plus GRPO/Liger `policy_loss` coverage.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit a77d8c3d94852af904be2a79f2d6f47d6a2e6f3f. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/experimental/ppo/ppo_trainer.py [hunk 1]
```diff
@@ -628,14 +630,31 @@ def repeat_generator():
...
+        # all-padding micro-batch has no statistic to report: writing the 0 that `masked_mean` returns for it would
+        # count as an observation and pull every mean toward 0, so its slot stays NaN and the `nanmean` reductions
+        # below leave it out.
+        approxkl_stats = torch.full(stats_shape, float("nan"), device=device)
+        pg_clipfrac_stats = torch.full(stats_shape, float("nan"), device=device)
+        pg_loss_stats = torch.full(stats_shape, float("nan"), device=device)
+        vf_loss_stats = torch.full(stats_shape, float("nan"), device=device)
+        vf_clipfrac_stats = torch.full(stats_shape, float("nan"), device=device)
+        entropy_stats = torch.full(stats_shape, float("nan"), device=device)
+        ratio_stats = torch.full(stats_shape, float("nan"), device=device)
```
[the remark this answers is not in the document]
Comment (behroozazarkhalili) [reply trl/experimental/ppo/ppo_trainer.py]: Confirmed. The buffers were allocated once, before the update loop, and the NaN-skip made that a regression: every slot used to be overwritten every update, so a single allocation was safe until slots could be skipped.

Fixed in 798885fb: the seven buffers are created at the top of each update. The regression test makes every row of the second update all padding, so that update has nothing to report and its `val/ratio` must be NaN; on the previous head it read the first update's value.
